alibi_detect/cd/pytorch/mmd_online.py

```diff
@@ -0,0 +1,191 @@
+            if self.device.type == 'cpu':
+                print('No GPU detected, fall back on CPU.')
+        else:
+            self.device = torch.device('cpu')
+
+        # initialize kernel
+        sigma = torch.from_numpy(sigma).to(self.device) if isinstance(sigma, np.ndarray) else None
+        self.kernel = kernel(sigma) if kernel == GaussianRBF else kernel
+
+        # compute kernel matrix for the reference data
+        self.x_ref = torch.from_numpy(self.x_ref).to(self.device)
+        self.k_xx = self.kernel(self.x_ref, self.x_ref, infer_sigma=(sigma is None))
+        self.infer_sigma = False
+
+        self._configure_thresholds()
+        self._initialise()
+
+    def _configure_ref_subset(self):
+        etw_size = 2*self.window_size-1  # etw = extended test window
+        rw_size = self.n - etw_size  # rw = ref-window
+        self.ref_inds = torch.randperm(self.n)[:rw_size]
+        self.k_xx_sub = self.k_xx[self.ref_inds][:, self.ref_inds]
+        self.k_xx_sub_sum = zero_diag(self.k_xx_sub).sum()/(rw_size*(rw_size-1))
+
+    def _configure_thresholds(self):
+
+        # Each bootstrap sample splits the reference samples into a sub-reference sample (x)
+        # and an extended test window (y). The extended test window will be treated as W overlapping
+        # test windows of size W (so 2W-1 test samples in total)
+
+        w_size = self.window_size
+        etw_size = 2*w_size-1  # etw = extended test window
+        rw_size = self.n - etw_size  # rw = sub-ref window
+
+        perms = [torch.randperm(self.n) for _ in range(self.n_bootstraps)]
+        x_inds_all = [perm[:-etw_size] for perm in perms]
+        y_inds_all = [perm[-etw_size:] for perm in perms]
+
+        print("Generating permutations of kernel matrix..")
+        # Need to compute mmd for each bs for each of W overlapping windows
+        # Most of the computation can be done once however
+        # We avoid summing the rw_size^2 submatrix for each bootstrap sample by instead computing the full
+        # sum once and then subtracting the relavent parts (k_xx_sum = k_full_sum - 2*k_xy_sum - k_yy_sum).
+        # We also reduce computation of k_xy_sum from O(nW) to O(W) by caching column sums
+
+        k_full_sum = zero_diag(self.k_xx).sum()
+        k_xy_col_sums_all = [
+            self.k_xx[x_inds][:, y_inds].sum(0) for x_inds, y_inds in
+            tqdm(zip(x_inds_all, y_inds_all), total=self.n_bootstraps)
+        ]
+        k_xx_sums_all = [(
+            k_full_sum - zero_diag(self.k_xx[y_inds][:, y_inds]).sum() - 2*k_xy_col_sums.sum()
+        )/(rw_size*(rw_size-1)) for y_inds, k_xy_col_sums in zip(y_inds_all, k_xy_col_sums_all)]
+        k_xy_col_sums_all = [k_xy_col_sums/(rw_size*w_size) for k_xy_col_sums in k_xy_col_sums_all]
+
+        # Now to iterate through the W overlapping windows
+        thresholds = []
+        for w in tqdm(range(w_size), "Computing thresholds"):
```

Comment:
Although I find progress bars by default helpful, in e.g. the outlier detectors we have a `verbose` flag and if set to *True* the progress bar during fitting of the detector becomes visible. @jklaise an opinion on this?

Reply:
Probably helps to be consistent. I believe keras fit has progress on by default, we could potentially do the same?

Reply:
I'll add a verbose flag. Config can take a while for large N and B so I think it's useful to be able to see progress.